Run funcbench via GKE

nevill opened the pull request:
xref #310 

it includes:
- [x] Run `funcbench` as a Kubernetes Job
- [x] Drop param `-i event.json` from `funcbench` cli, add `org / repo / branch / prNumber / regex / raceEnabled / postComment` as input parameters
- [ ] Replace docker images `nevill/prombench`, `nevill/funcbench` once the latest are released
- [ ] Let `funcbench` run on Sole-tenant nodes

Once it's merge, we also have to update Github actions in main repository

geekodour committed:
Added k8s jobs apply, wait and delete.

Signed-off-by: Hrishikesh Barman <hrishikeshbman@gmail.com>

nevill committed:
Kubernetes deployment for funcbench

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Use Job instead of Deployment to run funcbench test

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Set backoffLimit to 0 to make sure Job only run once

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Remove metadata from pod template

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
- Delete Job resource object when jobApply returns
- Add the way to test if job fails
- Increase wait time for method jobApply

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Make jobApply running infinitely

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Ignore the error of `jobDelete` when applying a Job resource

Signed-off-by: Nevill <nevill.dutt@gmail.com>

krasi-georgiev commented on pkg/provider/k8s/k8s.go:
```diff
@@ -179,6 +180,10 @@ func (c *K8s) ResourceApply(deployments []Resource) error {
 				err = c.statefulSetApply(resource)
+			case "job":
+				err = c.jobApply(resource)
+				// Even job deletion fails, it will be removed together with nodepool.
```
```suggestion
				// Ignore the error as even when job deletion fails, it will be removed together with nodepool.
```

krasi-georgiev commented on pkg/provider/k8s/k8s.go:
```diff
@@ -1282,6 +1354,37 @@ func (c *K8s) statefulSetReady(resource runtime.Object) (bool, error) {
+			return false, errors.Wrapf(err, "Checking Job resource:'%v' status failed err:%v", req.Name, err)
+		}
+
+		// current jobReady only works for non-parallel jobs
```
Please follow Golang commentary guidelines.
https://golang.org/doc/effective_go.html#commentary

krasi-georgiev commented on funcbench/manifests/benchmark/2_job.yaml:
```diff
@@ -0,0 +1,78 @@
+            value: {{ .LAST_COMMIT_SHA }}
+          - name: GOTMPDIR
+            value: /build/tmp
+          - name: TMPDIR
```
why do we need these 2 env vars? GOTMPDIR and TMPDIR

can we use the default linux tmp dir for all these?

also why do we need GITHUB_WORKSPACE?

nevill replied:
These 3 variables `GITHUB_WORKSPACE`, `GOTMPDIR`, `TMP_DIR` are used for generated testing binary, to make sure tests are run on the attached local SSD disk. But it can be replaced by specify disktype to `pd-ssd` when creating a VM. These will be removed later.

krasi-georgiev replied:
why not just use the linux tmp dir?
/tmp

nevill replied:
because `/tmp` is not on a SSD disk.

krasi-georgiev replied:
in this case this doesn't matter as there is no requirement for high io

krasi-georgiev commented on funcbench/manifests/benchmark/2_job.yaml:
```diff
@@ -0,0 +1,78 @@
+    }
+  BRANCH: {{ .BRANCH }}
+  RACE: {{ .RACE }}
+  REGEX: {{ .REGEX }}
```
what are these for?

nevill replied:
They are the inputs of `funcbench`.
See https://github.com/prometheus/prombench/blob/aad5c780d5cb2ef3e866f49f2b4a500416217bf1/funcbench/main.go#L231-L235
and 
https://github.com/prometheus/prombench/blob/aad5c780d5cb2ef3e866f49f2b4a500416217bf1/funcbench/main.go#L165

krasi-georgiev replied:
why not set them as env vars?

geekodour replied:
@nevill commentMonitor no longer works by sharing variables in files. To avoid confusion, I think it'll be best to replace these with command line arguments or environment variable whichever is more preferable here. I might be misunderstanding aswell, can you share the github actions workflow file used to trigger funcbench here?

krasi-georgiev replied:
yes actually I also meant command line variable

nevill replied:
@geekodour the workflow file is [here](https://github.com/nevill/prometheus/blob/master/.github/workflows/benchmark.yml)

krasi-georgiev replied:
this is to be changed to cli args right?

nevill replied:
right, will update this soon

krasi-georgiev commented on funcbench/manifests/benchmark/2_job.yaml:
```diff
@@ -0,0 +1,78 @@
+          mountPath: /build
+      containers:
+      - name: funcbench
+        image: nevill/funcbench:latest
```
we should use the official images `prombench/funcbench`

krasi-georgiev commented on funcbench/manifests/nodepools.yaml:
```diff
@@ -0,0 +1,15 @@
+  - name: funcbench-{{ .PR_NUMBER }}
+    initialnodecount: 1
+    config:
+      machinetype: n1-highmem-4
```
Is this the smalles machine that can be run as Sole-tenant nodes. If not please set to the lowest one possible to keep the cost of running these very low.

krasi-georgiev reviewed (commented):
Any changes required to the funcbnech cli?

nevill commented:
> Any changes required to the funcbnech cli?

No, there isn't any right now. But @geekodour should make sure the docker image is up to date.

nevill committed:
Update comments

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Update args of funcbench, based on the previous input params change

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Simplify the resource definitions

Signed-off-by: Nevill <nevill.dutt@gmail.com>

nevill committed:
Fix funcbench
- Use sh instead of bash for container environment
- Simplify working directory checking

nevill commented:
#343 will replace this one.
